core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -285,19 +288,23 @@ private void applyDeletedShards(final ClusterChangedEvent event) {
         if (routingNode == null) {
             return;
         }
-        Set<String> newShardAllocationIds = new HashSet<>();
+
+        final Map<Index, List<ShardRouting>> shardsByIndex = new HashMap<>();
```

Comment:
Why not go directly to Map<Index, Set<String>>? That we only need to build each set one time?